### ISSUE TITLE
Cannot round trip between Hall number and symmetry operations

### ISSUE THREAD

**Issue by kbarros** [spglib/spglib#132]
Hello,

Per this page, http://pmsl.planet.sci.kobe-u.ac.jp/~seto/?page_id=37&lang=en , Hall numbers 525 and 526 both correspond to the diamond-cubic space-group with HM symbol 'F d -3 m'.

Spglib allows getting symmetry from a Hall number, and then converting back to a Hall number:

```
sym = spglib.get_symmetry_from_database(526)
assert 525 == spglib.get_hall_number_from_symmetry(sym['rotations'], sym['translations'])
```

Interesting, this "round trip" is not the identity function, and takes 526 to 525. Is this expected behavior? It's not a blocker for my application, but I'm reporting it because the result seems surprising.

**Comment by atztogo** [comment]
> Is this expected behavior?

Yes.

**Comment by kbarros** [comment]
Thank you for the quick reply. I think this resolves the issue. I double checked the docs, and am not sure why `get_hall_number_from_symmetry` is "lossy" -- perhaps it would be useful to add a note?

**Comment by atztogo** [comment]
Can you define what you mean "lossy"?
Do you ask your question after understanding Hall symbol? http://cci.lbl.gov/sginfo/hall_symbols.html

**Comment by kbarros** [comment]
I know very little about space groups and Hall numbers, so thank you for the link, and maybe my suggestion was not a good one. By "lossy" I only meant to ask the following: if Hall numbers 525 and 526 can be mapped via the database to two different lists of symmetry operations S1 and S2, then why can't the lists of symmetry operations S1 and S2 map back to 525 and 526? If the answer is in the link you sent then that is sufficient. Feel free to close the issue.

**Comment by atztogo** [comment]
Because they are the choices of a space group type. The choice has to be made by users. Spglib returns the first one (from the top) appearing in the list of the Hall symbols corresponding to the space group type. http://spglib.github.io/spglib/api.html#spg-get-dataset-and-spg-get-dataset-with-hall-number

**Comment by afonari** [comment]
I think this is similar to https://github.com/spglib/spglib/issues/32.

To correctly assign hall number for a structure, one needs to compare operations obtained from get_symmetry_dataset with all hall groups with space group, and find one that is matching (get_hall_number_from_symmetry function).

I agree this seems weird.

**Comment by atztogo** [comment]
Please define more clearly the point you think weird so that I can understand. I assume you have read https://arxiv.org/abs/1808.01590 and understood it, e.g., difference between abstract group and its matrix representation in addition to the crystallographic convention. Then please raise a new issue with telling if this issue is about the design or the theory.

**Comment by lan496** [comment]
Let me get this discussion straight with a smaller example.

Consider space-group type P2 (No.3) and corresponding settings, Hall symbol P2y (3) and P2 (4):

```python
import spglib
import numpy as np
# Hall: P2y (3), ITA: P2 unique axis-b (No. 3)
sym_3_3 = spglib.get_symmetry_from_database(hall_number=3)
# Hall: P2 (4), ITA: P2 unique axis-c (No. 3)
sym_3_4 = spglib.get_symmetry_from_database(hall_number=4)
```

P2(4) is equivalent to P2y(3) up to orientation-preserving transformation, but their matrix representations are different:
```python
# ITA Sec. 1.5, Table 1.5.1.1
b_to_c = np.array([
    [0, 1, 0],
    [0, 0, 1],
    [1, 0, 0]
])

def transform(rotations, translations, P):
    new_rotations = [np.linalg.inv(P) @ R @ P for R in rotations]
    new_translations = [np.linalg.inv(P) @ t for t in translations]
    return new_rotations, new_translations

assert np.allclose(sym_3_4['rotations'], transform(sym_3_3['rotations'], sym_3_3['translations'], b_to_c)[0])
```

People seem to expect `get_hall_number_from_symmetry` (and `spg_get_hall_number_from_symmetry`) to return 4 for `sym_3_4`, but spglib returns 3 instead:
```python
assert spglib.get_hall_number_from_symmetry(sym_3_4['rotations'], sym_3_4['translations']) == 3
```

When an unconventional setting is used, people will expect `get_hall_number_from_symmetry` to raise error or return None. On the other hand, it also returns 3:
```python
# No Hall-symbol correspondence, but equivalent to P2
P_unconv = np.array([
    [1, 2, 0],
    [0, 1, 0],
    [0, 0, 1],
])
rotations_unconv, translations_unconv = transform(sym_3_3['rotations'], sym_3_3['translations'], P_unconv)

assert spglib.get_hall_number_from_symmetry(rotations_unconv, translations_unconv) == 3
```

I am not sure about C programming and spglib’s implementation, but the current behavior seems to come from the below code that only searches among ITA conventional settings (`spg_get_hall_number_from_symmetry` calls `spa_search_spacegroup_with_symmetry`):
https://github.com/spglib/spglib/blob/f8e6d5be6703466fa1444500a08d3fa623a0091d/src/spacegroup.c#L635-L640

**Comment by atztogo** [comment]
@lan496, I see your point. Then the problem is that the function name `get_hall_number_from_symmetry` is misleading. If what is expected is what as you wrote, then users can just use `get_symmetry_from_database` over all hall symbol indices and compare by themselves.

This function is implemented experimentally for uesr's request. I made it, but the user finally didn't reply. His request was looking for the space group type from a set of symmetry operations found by other code. This set can be non-conventional one. `search_spacegroup_with_symmetry` returns a hall symbol index, so probably I named the function name as it is without considering carefully.

I don't use this, and I don't know who use this. So this function may be removed in the next major release. This function was implemented very before. In those days, software dependence chain was not like right now. At that time I was not as strict as right now about the version number system, which was requested by some research who relied on spglib. Now I have to follow the semantic versioning, I can't remove or rename it easily until next major release.

**Comment by lan496** [comment]
> Then the problem is that the function name get_hall_number_from_symmetry is misleading.

Honestly, I also misunderstood the behavior of `get_hall_number_from_symmetry` just after reading the document.  Adding description in the document may decrease to mislead?

> I don't use this, and I don't know who use this. So this function may be removed in the next major release.

Some people seem to use this function to determine space-group type:
- abinit: https://github.com/abinit/abipy/blob/a89f721599ed78e820f6e84dae4ce27712c454fa/abipy/core/symmetries.py
- aiida-crystal17: https://github.com/aiidaplugins/aiida-crystal17/blob/c6642adc1636c3be968c5bed5ed7d4ccc389ac1c/aiida_crystal17/parsers/raw/parse_fort34.py
- aiida-gulp: https://github.com/aiidaplugins/aiida-gulp/blob/9bcd6636c3c97adac857296dd26b6e8757d37569/aiida_gulp/symmetry/symmetry.py
- aiida-crystal-dft: https://github.com/tilde-lab/aiida-crystal-dft/blob/625d626d829853a00d7ded11a8cdd7ab5b18a8c9/aiida_crystal_dft/io/f34.py

**Comment by atztogo** [comment]
Could it be possible for you to raise a PR for the documentation fix?

The function name should be changed in the next release, and currently one may be kept in the next major release but should raise DeprecationWarning. This is for me not to forget. I put the link to here from issue #65.

It seems people rely on spglib more than I think!

**Comment by lan496** [comment]
Sure! I will handle it.

**Comment by kbarros** [comment]
I am closing since the confusion seems to be resolved through documentation updates in #148. Please reopen if you think there is still something left to do.
